sympy/core/power.py

```diff
@@ -404,15 +404,17 @@ def _eval_Mod(self, q):
 
             '''
             For unevaluated Integer power, use built-in pow modular
-            exponentiation.
+            exponentiation, if powers are not too large wrt base.
             '''
             if self.base.is_Integer and self.exp.is_Integer and q.is_Integer:
                 b, e, m = int(self.base), int(self.exp), int(q)
-                # for very large powers
-                if m <= 10**18 and _log(e, 2) > 4*m**(1/4):
-                    # pollard rho for modulo will not be
-                    # slower than binary exponentiation
-                    
+                '''
+                For very large powers, use totient reduction.
+                Bound on m, is for safe factorization memory wise ie m^(1/4).
+                For pollard-rho to be faster than built-in pow and for e >= lg(m),
+                lg(e) > 4 m^(1/4) check is added.
```

Comment:
Should we also have `e >= totient(m)`?

Reply:
Actually `totient(m) >= lg(m)`, so `e >= lg(m)` is sufficient to add, I think.

Reply:
I think we should have better justification, or else add the condition.

Reply:
So, should I add the inequality `totient(m) >= lg(m)` to the documentation?

Reply:
I mean that the condition `e >= totient(m)` should be checked. I don't think that `e >= log(m)` would suffice.

Reply:
Oh, I think I misunderstood your initial comment.
`e >= lg(m)` suffices because `lg(m)` is the upper bound on the power of any prime in `m`.
It follows from the proof, that we only need this condition to be checked.

Reply:
I was thinking that if `e < totient(m)`, then it would be reasonable to compute `pow(b, e, m)` instead of `pow(b, phi + e%phi, m)` where the exponent is bigger than `e`.

Reply:
Yes, fair point. But unfortunately `totient(m)` itself takes `O(m^(1/4))` which can be slow to compute inside an if condition, we would be better off using `pow(b,e,m)` directly in this case.

Reply:
Maybe it could be checked in the `if` block after `phi` has been computed. That block could redefine `e` if it is bigger than `phi` and then there would be a single `return pow(b, e, m)` at the end.

Reply:
Yes, that could be done but by the virtue of entering the `if` block, we get `lg(e) > 4 m^(1/4)`, which is equivalent to saying, `e > 2^(4 m^(1/4))` which I think would always be bigger than `totient(m)` or `m` for that matter (as `totient(m) <= m`).

Reply:
Yes, it really looks like `2^(4 m^(1/4))` is much bigger than `m`. (I didn't prove it but I think that it wouldn't be hard.) So I agree that checking `e >= totient(m)`  will not be needed.

Reply:
Yes, the proof is easy just take `log base m` for `m!=1`, and we get the result.